### PR TITLE
Add extraction progress output to GetFSFromImage/GetFSFromLayers

### PR DESCRIPTION
**Pull request by mafredri**
This PR allows printing extraction progress during unpack. Since we can't access the underlying reader for the compressed data, we can only print accurate progress at every layer change. Instead we print `...` every 1 second (triggered via read) to indicate that the extraction is still in progress.

Fixes https://github.com/coder/envbuilder/issues/124

```
#2: Building stage 'mcr.microsoft.com/devcontainers/typescript-node:18-bookworm' [idx: '1', base-idx: '-1']
#2: Unpacking rootfs as cmd ADD install-vscode.sh /root/ requires it.
#2: Extracting image layers to /
#2: Extracting layer 1/22 (0.0%)
#2: Extracting layer 1/22 (0.0%) .
#2: Extracting layer 1/22 (0.0%) ..
#2: Extracting layer 2/22 (7.6%)
#2: Extracting layer 3/22 (11.3%)
#2: Extracting layer 3/22 (11.3%) .
#2: Extracting layer 3/22 (11.3%) ..
#2: Extracting layer 4/22 (21.1%)
#2: Extracting layer 4/22 (21.1%) .
#2: Extracting layer 4/22 (21.1%) ..
#2: Extracting layer 4/22 (21.1%) ...
#2: Extracting layer 4/22 (21.1%) ....
#2: Extracting layer 4/22 (21.1%) .....
#2: Extracting layer 4/22 (21.1%) ......
#2: Extracting layer 5/22 (53.4%)
#2: Extracting layer 6/22 (53.4%)
#2: Extracting layer 6/22 (53.4%) .
#2: Extracting layer 7/22 (60.4%)
#2: Extracting layer 8/22 (60.6%)
#2: Extracting layer 9/22 (60.6%)
#2: Extracting layer 10/22 (61.5%)
#2: Extracting layer 11/22 (61.5%)
#2: Extracting layer 12/22 (61.5%)
#2: Extracting layer 13/22 (61.5%)
#2: Extracting layer 14/22 (61.5%)
#2: Extracting layer 14/22 (61.5%) .
#2: Extracting layer 14/22 (61.5%) ..
#2: Extracting layer 15/22 (72.1%)
#2: Extracting layer 15/22 (72.1%) .
#2: Extracting layer 16/22 (82.0%)
#2: Extracting layer 17/22 (83.6%)
#2: Extracting layer 17/22 (83.6%) .
#2: Extracting layer 18/22 (90.5%)
#2: Extracting layer 19/22 (90.5%)
#2: Extracting layer 20/22 (90.5%)
#2: Extracting layer 21/22 (90.5%)
#2: Extracting layer 22/22 (90.5%)
#2: Extracting layer 22/22 (90.5%) .
#2: Extraction complete
#2: Using files from context: [/workspaces/vscode/.devcontainer/install-vscode.sh]
```

### CODE REVIEW

**Comment by mafredri** [comment]
Whoops, just noticed this:
```
#2: Extracting image layers to / (0 bytes)
#2: Extraction complete 0/0 (-9223372036854775808%)
```
I'll push a fix later.

File: pkg/util/fs_util.go [hunk 1]
```diff
@@ -225,10 +262,38 @@ func GetFSFromLayers(root string, layers []v1.Layer, opts ...FSOpt) ([]string, e
 
 			extractedFiles = append(extractedFiles, filepath.Join(root, cleanedName))
 		}
+
+		extractedBytes += layerSizes[i]
 	}
+
+	if cfg.printExtractionProgress {
+		logrus.Infof("Extraction complete %d/%d (%d%%)", extractedBytes, totalSize, int(math.Round(float64(extractedBytes)/float64(totalSize)*100)))
```
Comment (johnstcn) [inline pkg/util/fs_util.go:270]: Why not just use `%0.2f` and let fmt do the rounding?

File: pkg/util/fs_util.go [hunk 2]
```diff
@@ -163,20 +170,50 @@ func GetFSFromLayers(root string, layers []v1.Layer, opts ...FSOpt) ([]string, e
...
 
+	var totalSize int64
+	var layerSizes []int64
+	for _, l := range layers {
+		layerSize, err := l.Size()
+		if err != nil {
+			return nil, err
+		}
+		layerSizes = append(layerSizes, layerSize)
+		totalSize += layerSize
```
Comment (johnstcn) [inline pkg/util/fs_util.go:181]: nit: maybe check that layerSize > 0?

File: pkg/util/fs_util.go [hunk 3]
```diff
@@ -225,10 +262,38 @@ func GetFSFromLayers(root string, layers []v1.Layer, opts ...FSOpt) ([]string, e
...
+	io.ReadCloser
+	n     int64
+	t     time.Time
+	after time.Duration
+	print func(int64)
+}
+
+func (r *printAfterReader) Read(p []byte) (n int, err error) {
+	n, err = r.ReadCloser.Read(p)
+	r.n += int64(n)
```
Comment (johnstcn) [inline pkg/util/fs_util.go:286]: nit: check n > 0

File: pkg/util/fs_util.go [hunk 4]
```diff
@@ -126,6 +127,12 @@ func IncludeWhiteout() FSOpt {
 	}
 }
 
+func PrintExtractionProgress() FSOpt {
```
Comment (johnstcn) [inline pkg/util/fs_util.go:130]: func opts are the 🐐

**Review by johnstcn — APPROVED** [review]
Some nits, but LGTM otherwise. I don't need to review again 👍

File: pkg/util/fs_util.go [hunk 5]
```diff
@@ -163,20 +170,50 @@ func GetFSFromLayers(root string, layers []v1.Layer, opts ...FSOpt) ([]string, e
...
+	var extractedBytes int64
 	for i, l := range layers {
 		if mediaType, err := l.MediaType(); err == nil {
 			logrus.Tracef("Extracting layer %d of media type %s", i, mediaType)
 		} else {
 			logrus.Tracef("Extracting layer %d", i)
 		}
 
+		if cfg.printExtractionProgress {
+			logrus.Infof("Extracting layer %d %d/%d (%d%%)", i, extractedBytes, totalSize, int(math.Round(float64(extractedBytes)/float64(totalSize)*100)))
```
Comment (dannykopping) [inline pkg/util/fs_util.go:198]: Showing how many layers there are will also be helpful.

```suggestion
			logrus.Infof("Extracting layer %d of %d %d/%d (%d%%)", i, len(layers), extractedBytes, totalSize, int(math.Round(float64(extractedBytes)/float64(totalSize)*100)))
```

Comment (dannykopping) [inline pkg/util/fs_util.go:198]: Be careful about div by 0 here

File: pkg/util/fs_util.go [hunk 6]
```diff
@@ -163,20 +170,50 @@ func GetFSFromLayers(root string, layers []v1.Layer, opts ...FSOpt) ([]string, e
 		return nil, errors.New("must supply an extract function")
 	}
 
+	var totalSize int64
```
Comment (dannykopping) [inline pkg/util/fs_util.go:173]: IMHO showing the raw number of bytes is not super useful for most folks; can we summarise these values into MB perhaps?

File: pkg/util/fs_util.go (hunk 5 above)
Comment (dannykopping) [inline pkg/util/fs_util.go:198]: Nit: I'd suggest adding some tabs here so the output stays aligned

File: pkg/util/fs_util.go [hunk 7]
```diff
@@ -163,20 +170,50 @@ func GetFSFromLayers(root string, layers []v1.Layer, opts ...FSOpt) ([]string, e
...
+	var layerSizes []int64
+	for _, l := range layers {
+		layerSize, err := l.Size()
+		if err != nil {
+			return nil, err
```
Comment (dannykopping) [inline pkg/util/fs_util.go:178]: Nit: please wrap err

File: pkg/util/fs_util.go [hunk 8]
```diff
@@ -163,20 +170,50 @@ func GetFSFromLayers(root string, layers []v1.Layer, opts ...FSOpt) ([]string, e
...
 			return nil, err
 		}
 		defer r.Close()
 
+		if cfg.printExtractionProgress {
+			r = &printAfterReader{
+				ReadCloser: r,
+				after:      time.Second,
+				print: func(int64) {
+					logrus.Infof("Extracting layer %d...", i)
```
Comment (dannykopping) [inline pkg/util/fs_util.go:212]: As an aside: I think it'd be even more useful to have a timestamp at the start of every log line so we could see how long things took after the fact.

File: pkg/util/fs_util.go [hunk 9]
```diff
@@ -225,10 +262,38 @@ func GetFSFromLayers(root string, layers []v1.Layer, opts ...FSOpt) ([]string, e
...
+func (r *printAfterReader) Read(p []byte) (n int, err error) {
+	n, err = r.ReadCloser.Read(p)
+	r.n += int64(n)
+	if r.t.IsZero() {
+		r.t = time.Now()
+	}
+	if time.Since(r.t) >= r.after {
+		r.print(r.n)
+		r.t = time.Now()
+	}
```
Comment (dannykopping) [inline pkg/util/fs_util.go:293]: Nit: I think the more idiomatic approach here would be spawning a goroutine with a `time.Ticker`.

**Review by dannykopping — APPROVED** [review]
Overall LGTM

Left a couple nits & suggestions

Q: did you consider passing a `io.TeeReader` to the extraction function? That might give you insight into the progress when `io.Copy` occurs [here](https://github.com/coder/kaniko/blob/main/pkg/util/fs_util.go#L302-L347) (I haven't validated this).

**Review by mtojek — APPROVED** [review]
👍

File: pkg/util/fs_util.go (hunk 9 above)
Comment (mafredri) [reply pkg/util/fs_util.go:293]: Hmm, I see what you mean. I didn't really want to spawn a goroutine for this that needs to be stopped, but I could refactor this in a different way that makes more sense with the ticker. Then we can keep printing even when no new reads happen, but we would then have to rely on close being called.

File: pkg/util/fs_util.go (hunk 8 above)
Comment (mafredri) [reply pkg/util/fs_util.go:212]: We could consider enabling that on the logger lever, but I don't think we should diverge for these log lines. Usually timestamps would be available e.g. via container logs or similar. Not against showing an elapsed human readable time, but then again I want to keep this simple 😄.

File: pkg/util/fs_util.go (hunk 6 above)
Comment (mafredri) [reply pkg/util/fs_util.go:173]: We could just remove it entirely and show percentage, to keep it simple? In my experience this type of output quite commonly shows bytes, which can make parsing easier (if anyone's into that).

File: pkg/util/fs_util.go (hunk 5 above)
Comment (mafredri) [reply pkg/util/fs_util.go:198]: I agree, but I'm on the fence about diverging from what Kaniko did beforehand. 🤔

File: pkg/util/fs_util.go (hunk 6 above)
Comment (mafredri) [reply pkg/util/fs_util.go:173]: Thoughts?
```
#2: Unpacking rootfs as cmd ADD install-vscode.sh /root/ requires it.
#2: Extracting image layers to /
#2: Extracting layer 0 (0%)
#2: Extracting layer 0...
#2: Extracting layer 0...
#2: Extracting layer 1 (8%)
#2: Extracting layer 1...
#2: Extracting layer 2 (11%)
#2: Extracting layer 2...
#2: Extracting layer 2...
#2: Extracting layer 2...
#2: Extracting layer 3 (21%)
#2: Extracting layer 3...
#2: Extracting layer 3...
#2: Extracting layer 3...
#2: Extracting layer 3...
#2: Extracting layer 3...
#2: Extracting layer 3...
#2: Extracting layer 3...
#2: Extracting layer 3...
#2: Extracting layer 3...
#2: Extracting layer 4 (53%)
#2: Extracting layer 5 (53%)
#2: Extracting layer 5...
#2: Extracting layer 5...
#2: Extracting layer 6 (60%)
```

File: pkg/util/fs_util.go (hunk 2 above)
Comment (mafredri) [reply pkg/util/fs_util.go:181]: I don't see any need for the check here, 0 doesn't add anything to the total.

File: pkg/util/fs_util.go (hunk 6 above)
Comment (dannykopping) [reply pkg/util/fs_util.go:173]: Looks great!

File: pkg/util/fs_util.go (hunk 9 above)
Comment (dannykopping) [reply pkg/util/fs_util.go]: Don't stress, not worth the refactor right now

**Comment by dannykopping** [comment]
> > Q: did you consider passing a `io.TeeReader` to the extraction function? That might give you insight into the progress when `io.Copy` occurs [here](https://github.com/coder/kaniko/blob/main/pkg/util/fs_util.go?rgh-link-date=2024-05-23T07%3A27%3A16Z#L302-L347) (I haven't validated this).
> 
> I'm not following what this would achieve? We're already monitoring reads via the printing reader and writes we can't know anything about.

In `func (r *printAfterReader) Read(p []byte) (n int, err error) {` you're printing `.` while the extracting is in progress, but I'm saying you could monitoring the extraction of each layer itself (I think). In any case, this is great as it is - just an idea for later if we have some big layers which take ages to extract and we need more fine-grained insight into that process.

**Comment by mafredri** [comment]
> In `func (r *printAfterReader) Read(p []byte) (n int, err error) {` you're printing `.` while the extracting is in progress, but I'm saying you could monitoring the extraction of each layer itself (I think). In any case, this is great as it is - just an idea for later if we have some big layers which take ages to extract and we need more fine-grained insight into that process.

@dannykopping Ah, I see. This is something I wanted to do, but as I mentioned in the PR description, it's not possible. We only know the _compressed_ size and the number of _uncompressed_ bytes read. 😔 Correlating these will always be inaccurate.

**Pull request merged**